### PR TITLE
Non training Batch Norm operator has bad performance for it running into tensorflow's non fused batch norm API

### PR DESCRIPTION
**Pull request by zwang36wang**
In current Keras code with Tensorflow as it’s backend, the non-training batch norm operator/layer will finally run into Tensorflow’s old non-fused batch norm API (tf.nn.batch_normalization), which result in bad performance in both CPU and GPU.

http://github.com/keras-team/keras/issues/10058

Test result:
 test env: with Tensorflow(commit a543d9471047ca3f6881c87105fcbe2cdff9207d Date:   Thu May 10 17:43:30 2018, local build), python3.4, centos7.4
 test cases:
      "pytest  ./tests/keras/layers/normalization_test.py"  <all passed>
      "pytest  ./tests"      <keep same result as without this PR's modification>

### CODE REVIEW

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -1866,6 +1866,26 @@ def batch_normalization(x, mean, var, beta, gamma, epsilon=1e-3):
     # Returns
         A tensor.
     """
+    if ndim(x) == 4:
+        # The CPU implementation of FusedBatchNorm only support NHWC
+        if len(mean.shape) == 1:
+            tf_data_format = 'NHWC'
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: This is incorrect. What data format to use depends on what `axis` argument was passed to the BN layer.

File: keras/backend/tensorflow_backend.py [hunk 2]
```diff
@@ -1866,6 +1866,26 @@ def batch_normalization(x, mean, var, beta, gamma, epsilon=1e-3):
     # Returns
         A tensor.
     """
+    if ndim(x) == 4:
+        # The CPU implementation of FusedBatchNorm only support NHWC
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: NCHW should be supported on GPU since it offers better performance. For conv ops we have a test checking whether we run on GPU or not, check it out

File: keras/backend/tensorflow_backend.py [hunk 3]
```diff
@@ -1866,6 +1866,26 @@ def batch_normalization(x, mean, var, beta, gamma, epsilon=1e-3):
...
+        if len(mean.shape) == 1:
+            tf_data_format = 'NHWC'
+        else:
+            tf_data_format = None
+        if (tf_data_format is not None and
+                beta is not None and gamma is not None):
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: If beta/gamma are None, we should create `ones` vectors for them so we can use fused ops anyway.

File: keras/backend/tensorflow_backend.py (hunk 1 above)
Comment (zwang36wang) [reply keras/backend/tensorflow_backend.py]: Agree. It's much more clean way to introduce the axis/reduction_axes parameter for backend interface function "batch_normalization" just as the "normalize_batch_in_training" function, then all the three backend cntk/theano/tensorflow will have axis info in batch_normalization function, though current theano/cntk don't need to deal axis info( I don't know whether they have fused BatchNorm version imp).

Previous I don't want to change the "batch_normalization" backend interface, and tried to get axis info from shape, the logic is little tricky there.

File: keras/backend/tensorflow_backend.py (hunk 2 above)
Comment (zwang36wang) [reply keras/backend/tensorflow_backend.py]: Agree

File: keras/backend/tensorflow_backend.py (hunk 3 above)
Comment (zwang36wang) [reply keras/backend/tensorflow_backend.py]: Agree to extend.  
 If beta is None, let's to be zeros.
 If gamma is None, let's to be ones.

File: keras/backend/cntk_backend.py [hunk 4]
```diff
@@ -1062,7 +1062,7 @@ def _moments(x, axes=None, shift=None, keep_dims=False):
     return mean, variance
 
 
-def batch_normalization(x, mean, var, beta, gamma, epsilon=1e-3):
+def batch_normalization(x, mean, var, beta, gamma, axis, epsilon=1e-3):
```
[the remark this answers is not in the document]
Comment (zwang36wang) [reply keras/backend/cntk_backend.py]: Agree

File: keras/backend/tensorflow_backend.py [hunk 5]
```diff
@@ -1849,7 +1849,7 @@ def normalize_batch_in_training(x, gamma, beta,
                                                           epsilon=epsilon)
 
 
-def batch_normalization(x, mean, var, beta, gamma, epsilon=1e-3):
+def batch_normalization(x, mean, var, beta, gamma, axis, epsilon=1e-3):
```
[the remark this answers is not in the document]
Comment (zwang36wang) [reply keras/backend/tensorflow_backend.py]: Agree.

File: keras/backend/tensorflow_backend.py [hunk 6]
```diff
@@ -1861,11 +1861,39 @@ def batch_normalization(x, mean, var, beta, gamma, epsilon=1e-3):
         var: Variance of batch.
         beta: Tensor with which to center the input.
         gamma: Tensor by which to scale the input.
+        axis: Integer, the axis that should be normalized.
+                (typically the features axis).
```
[the remark this answers is not in the document]
Comment (zwang36wang) [reply keras/backend/tensorflow_backend.py]: Ok.

File: keras/backend/theano_backend.py [hunk 7]
```diff
@@ -730,7 +730,7 @@ def normalize_batch_in_training(x, gamma, beta,
     return normed, mean, T.inv(stdinv ** 2)
 
 
-def batch_normalization(x, mean, var, beta, gamma, epsilon=1e-3):
+def batch_normalization(x, mean, var, beta, gamma, axis, epsilon=1e-3):
```
[the remark this answers is not in the document]
Comment (zwang36wang) [reply keras/backend/theano_backend.py]: Ok

File: keras/layers/normalization.py [hunk 8]
```diff
@@ -169,6 +170,7 @@ def normalize_inference():
                     self.moving_variance,
                     self.beta,
                     self.gamma,
+                    self.axis,
```
[the remark this answers is not in the document]
Comment (zwang36wang) [reply keras/layers/normalization.py]: Ok

File: keras/backend/tensorflow_backend.py [hunk 9]
```diff
@@ -1861,11 +1861,39 @@ def batch_normalization(x, mean, var, beta, gamma, epsilon=1e-3):
...
+    if ndim(x) == 4:
+        # The CPU implementation of FusedBatchNorm only support NHWC
+        if axis == 1:
+            tf_data_format = 'NCHW'
+        elif axis == 3:
+            tf_data_format = 'NHWC'
+        else:
+            tf_data_format = None
+
+        if ((tf_data_format == 'NHWC') or (tf_data_format == 'NCHW' and _has_nchw_support())):
```
[the remark this answers is not in the document]
Comment (zwang36wang) [reply keras/backend/tensorflow_backend.py]: Cool, much more clean code.
